src/libstd/collections/hash/map.rs

```diff
@@ -1370,6 +1370,37 @@ impl<K, V, S> Eq for HashMap<K, V, S>
 {
 }
 
+#[unstable(feature = "hashmap_hash", issue = "0")]
+impl<K, V, S> Hash for HashMap<K, V, S>
+    where K: Eq + Hash,
+          V: Hash,
+          S: BuildHasher
+{
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        // We must preserve:  x == y -> hash(x) == hash(y).
+        // HashMaps have no order, so we must use a commutative operation
+        // (.wrapping_add) so that the order does not matter.
+        //
+        // For this to hold, we must also ensure that the hashing of
+        // individual elements does not depend on the state of the given Hasher.
+        // So we ensure that hashing each individual element starts with the
+        // same state.
+        //
+        // Unfortunately, we can't .clone() the hasher since we can't add more
+        // constraints than H being a Hasher. With some sort of ConstraintKinds
+        // we might be able do so in the future.
+        hasher.write_u64(
+            self.iter()
+                .map(|(k, v)| {
```

Comment:
no need to destructure and re-tuple below.

Reply:
Good point, fixing that.